addons/payment/models/payment_provider.py

```diff
@@ -487,58 +493,85 @@ def _get_compatible_providers(
             *self.env['payment.provider']._check_company_domain(company_id),
             ('state', 'in', ['enabled', 'test']),
         ]
+        providers = self.env['payment.provider'].search(domain)
 
-        # Handle the is_published state.
-        if not self.env.user._is_internal():
-            domain = expression.AND([domain, [('is_published', '=', True)]])
+        # Add payment providers to the report and set the defaults.
+        payment_utils.add_to_report(report, providers)
 
         # Handle the partner country; allow all countries if the list is empty.
         partner = self.env['res.partner'].browse(partner_id)
         if partner.country_id:  # The partner country must either not be set or be supported.
-            domain = expression.AND([
-                domain, [
-                    '|',
-                    ('available_country_ids', '=', False),
-                    ('available_country_ids', 'in', [partner.country_id.id]),
-                ]
-            ])
+            unfiltered_providers = providers
+            providers = providers.filtered(
+                lambda p: not p.available_country_ids
+                          or partner.country_id.id in p.available_country_ids.ids
+            )
+            payment_utils.add_to_report(
+                report,
+                unfiltered_providers - providers,
+                available=False,
+                reason=REPORT_REASONS_MAP['incompatible_country'],
+            )
 
         # Handle the maximum amount.
         currency = self.env['res.currency'].browse(currency_id).exists()
         if not is_validation and currency:  # The currency is required to convert the amount.
             company = self.env['res.company'].browse(company_id).exists()
             date = fields.Date.context_today(self)
             converted_amount = currency._convert(amount, company.currency_id, company, date)
-            domain = expression.AND([
-                domain, [
-                    '|', '|',
-                    ('maximum_amount', '>=', converted_amount),
-                    ('maximum_amount', '=', False),
-                    ('maximum_amount', '=', 0.),
-                ]
-            ])
+            unfiltered_providers = providers
+            providers = providers.filtered(
+                lambda p: not p.maximum_amount
+                          or p.maximum_amount >= converted_amount
```

Comment:
Hm, might be interesting to use `float_compare` here, to avoid (very rare, I know) problems in the future? :thinking:  Not sure though, should be good like this too.

Reply:
sure, but which rounding should we take?